src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,369 @@
+      uint64_t check_channel_status_interval_milliseconds,
+      uint64_t server_unavailable_timeout_seconds,
+      std::function<void()> server_unavailable_timeout_callback,
+      const std::string &server_name) {
+    auto retryable_grpc_client = std::shared_ptr<RetryableGrpcClient>(
+        new RetryableGrpcClient(channel,
+                                io_context,
+                                max_pending_requests_bytes,
+                                check_channel_status_interval_milliseconds,
+                                server_unavailable_timeout_seconds,
+                                server_unavailable_timeout_callback,
+                                server_name));
+    return retryable_grpc_client;
+  }
+
+  RetryableGrpcClient(const RetryableGrpcClient &) = delete;
+  RetryableGrpcClient &operator=(const RetryableGrpcClient &) = delete;
+
+  template <typename Service, typename Request, typename Reply>
+  void CallMethod(PrepareAsyncFunction<Service, Request, Reply> prepare_async_function,
+                  GrpcClient<Service> &grpc_client,
+                  const std::string &call_name,
+                  const Request &request,
+                  const ClientCallback<Reply> &callback,
+                  const int64_t timeout_ms) {
+    auto executor = new Executor(
+        [callback](const ray::Status &status) { callback(status, Reply()); });
+    std::weak_ptr<RetryableGrpcClient> weak_self = shared_from_this();
+    auto operation_callback = [weak_self, request, callback, executor, timeout_ms](
+                                  const ray::Status &status, Reply &&reply) {
+      auto self = weak_self.lock();
+      if (status.ok() || !IsGrpcRetryableStatus(status) || !self) {
+        callback(status, std::move(reply));
+        delete executor;
+      } else {
+        // In case of transient network error, we queue the request and these requests
+        // will be executed once network is recovered.
+        auto request_bytes = request.ByteSizeLong();
+        if (self->pending_requests_bytes_ + request_bytes >
+            self->max_pending_requests_bytes_) {
+          RAY_LOG(WARNING)
+              << "Pending queue for failed request has reached the "
+              << "limit. Blocking the current thread until network is recovered";
+          if (!self->server_unavailable_timeout_time_.has_value()) {
+            self->server_unavailable_timeout_time_ =
+                absl::Now() + absl::Seconds(self->server_unavailable_timeout_seconds_);
+          }
+          while (self->server_unavailable_timeout_time_.has_value() && !self->shutdown_) {
+            self->CheckChannelStatus(false);
+            std::this_thread::sleep_for(std::chrono::milliseconds(
+                self->check_channel_status_interval_milliseconds_));
+          }
+          if (self->shutdown_) {
+            callback(Status::Disconnected("GRPC client has been shutdown."), Reply());
+            delete executor;
```

Comment:
yeah as I said, please avoid manually `delete` and `new`; it's ridiculously easy to forget release somewhere.

Reply:
I will try to write a scoped guard, which might help here.

Reply:
Checkout: https://github.com/ray-project/ray/pull/48622